Fixing another `shared_memory` leak in `raster_calculator`.

phargogh opened the pull request:
This PR fixes another shared memory leak in `raster_calculator` found during the NCI work.

Fixes #247 
Fixes #309

phargogh committed:
Found another place where shared memory was leaking.

RE:#247

phargogh committed:
Noting shared memory leak fix in HISTORY. RE:#247

phargogh committed:
Updating how the version information is read.

This is in keeping with the latest recommendations from setuptools_scm.

phargogh committed:
Removing the environment specifier.

phargogh committed:
Removing the deprecated pkg_resources import.

davemfish commented on requirements.txt:
```diff
@@ -9,3 +9,4 @@ numpy>=1.10.1
 Rtree>=0.8.3
 scipy>=0.14.1,!=0.19.1
 Shapely>=1.6.4
+importlib_metadata  # technically only required on python < 3.8; easier to install with conda across all versions
```
Looks like this is not working in the `3.7` envs. But the failure is actually during environment creation, so maybe some other library is using this package and no longer getting the version it expected?

phargogh replied:
Yeah, this turned out to be an apparent incompatibility between the `build` and `importlib_metadata` packages that is resolved by upgrading to the latest version of `build`, which is now distributed on `conda-forge` as `python-build`.

davemfish commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -589,13 +589,14 @@ def raster_calculator(
+                shared_memory.close()
+                shared_memory.unlink()
+                LOGGER.debug(
+                    f'unlinked shared memory for process {os.getpid()}')
```
Ahhh, whitespace. Github's side-by-side diff not doing me any favors here.

phargogh replied:
Yeah, it's easy for the diff to hide this!  Isn't there even a whitespace-ignoring view?  So if the only changes were whitespace, you won't see the changes at all?

davemfish reviewed (changes requested):
Hey @phargogh , the change makes sense to me. But there's some sort of library incompatibility going in with python `3.7`

phargogh committed:
Trying to run with mamba instead of conda.

Also updating the build package to python-build.

This should overall improve build speeds, but also get us later fixes
and features from the build package, which should also address the more
immediate build problem. RE:#247 RE:#309

phargogh committed:
Trying a different method of writing out requirements.

RE:#247 RE:#309

phargogh committed:
I forgot the writeable bit.

RE:#247 RE:#309

phargogh committed:
Using pip-compatible requirements file. RE:#247 #309

phargogh committed:
Trying a different method of specifying python version.

RE:#247 #309

phargogh committed:
Trying to build just with requirements. RE:#247 #309

phargogh committed:
Maybe the issue is with an empty line?

RE:#247 #309

phargogh committed:
Restoring extra specs for all-at-once deps resolution.

RE:#247 #309

phargogh committed:
Moving flake8 and pytest to the package installation step.

RE:#247 #309

phargogh commented:
Thanks for pointing out the build failure there, @davemfish ! This turned out to be an unfortunate combination of an old version of `build` being installed and `setuptools` on its way out.  I upgraded to the latest version of the `build` package (distributed now on `conda-forge` as `python-build`) and also took this opportunity to roughly halve the GHA runtime by using `mamba` for environment management as we do in InVEST.  The python 3.7 tests appear to be passing so far, fingers crossed that it'll keep going for the rest of the tests.

This also means that this PR addresses #309

davemfish reviewed (approved):
Thanks for figuring all that out!
